### ISSUE TITLE
[GSoC] TFDS should fail when user try to generate an old version

### ISSUE THREAD

**Issue by Conchylicultor** [tensorflow/datasets#1708]
TFDS only support generation of dataset for the last versions.

We should make sure to raise an error here if the user try to generate an older version. The `download_and_prepare()` function should be updated here to raise an error: https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/dataset_builder.py#L290

A version can only be installed if the version is the `self.canonical_version` or the `max(self.versions)`

For example:
```
VERSION = tfds.core.Version('1.0.0')  # Can be generated
SUPPORTED_VERSION = [
    tfds.core.Version('2.1.0'),  # Can be generated
    tfds.core.Version('2.0.0'),  # Cannot be generated
    tfds.core.Version('0.1.0'),  # Cannot be generated
]
```

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  I am working on it, send you pr soon, We have to allow only these two `self.canonical_version` or the `max(self.versions)` ?

**Comment by Eshan-Agarwal** [comment]
Also can you please give a name of dataset which have these versions for testing purpose

**Comment by Conchylicultor** [comment]
Thank you for looking into this.

> We have to allow only these two self.canonical_version or the max(self.versions) ?

@Eshan-Agarwal yes, this is correct.

> Also can you please give a name of dataset which have these versions for testing purpose

For testing, you should add a unittest in `dataset_builder_test.py` to make sure this works properly.

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  thanks for your quick response, working on it

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  in the link above by replacing `self.versions` with `self.canonical_version or  max(self.versions)`  can work ?

**Comment by preeti13456** [comment]
Interested

**Comment by preeti13456** [comment]
https://github.com/preeti13456/Projects-and-research/blob/preeti13456-patch-1/DatasetBuilder.ipynb

**Comment by vijayphoenix** [comment]
This can be closed 0e1e4f8
